xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -527,14 +527,18 @@ private boolean retainCluster(String cluster) {
 
     private void releaseCluster(final String cluster) {
       int count = clusterRefs.get(cluster).refCount.decrementAndGet();
+      if (count < 0) {
+        throw new AssertionError();
+      }
       if (count == 0) {
         syncContext.execute(new Runnable() {
           @Override
           public void run() {
-            if (clusterRefs.get(cluster).refCount.get() == 0) {
-              clusterRefs.remove(cluster);
-              updateResolutionResult();
+            if (clusterRefs.get(cluster).refCount.get() != 0) {
```

Comment:
Since this is being added to the end of the queue, it is possible that another update has come in restoring use of the cluster before this runs, in which case doing nothing is appropriate.

Reply:
Because there's a refcount _and_ the clusterRefs map, the threading simply breaks down if the refcount goes from 0 → 1. That's why we already don't use an atomic increment, and instead use compareAndSet to increase the value:
https://github.com/grpc/grpc-java/blob/ca4819ac6dbd3d99a1f20c1e08625020b8a48994/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L523-L528

1 → 0 is a one-way state transition, and the only option is to throw away the object.

In my debugging, I was worried that there could have been a bug in the reference counting that left things in a corrupted state. I don't think there is any such bug, but I'd much rather see an exception than silent state corruption.